File: puzzles/The-Fall-Episode-1/the_fall_episode_1.dart
import 'dart:io';
import 'dart:math';


String readLineSync() {
  String? s = stdin.readLineSync();
  return s == null ? '' : s;
}

void logd(String aStr, [bool aNextLine=true]) {
  stderr.write(aStr+(aNextLine ? "\n" : ""));
}

class Direction {
  static final top    = Direction("TOP",    0,  -1);
  static final right  = Direction("RIGHT",  1, 0);
  static final bottom = Direction("BOTTOM", 0,  1);
  static final left   = Direction("LEFT",   -1, 0);
  
  Direction(this._str, this._dx, this._dy);

  String  get str => _str;
  int     get dx  => _dx;
  int     get dy  => _dy;

  final String _str;
  final int    _dx;
  final int    _dy;

  Direction get from {
    if (Direction.top == this ) { return Direction.bottom; }
    else if (Direction.right == this) { return Direction.left; }
    else if (Direction.bottom == this) { return Direction.top; }
    else if (Direction.left == this) { return Direction.right; }
    else { throw FormatException("Invalid direction");}
  }

  static Direction read(String aS) {
    switch(aS) {
      case "TOP"    : return Direction.top;
      case "RIGHT"  : return Direction.right;
      case "BOTTOM" : return Direction.bottom;
      case "LEFT"   : return Direction.left;
      default       : throw FormatException("Invalid direction: ${aS}");
    }
  }
}


class RoomInOut {
  final List<Direction> _ins;
  final List<Direction> _outs;

  List<Direction> get ins   => _ins;
  List<Direction> get outs  => _outs;

  RoomInOut(List<Direction> this._ins, List<Direction> this._outs) {}
}

typedef RoomType = int;

final ROOM_TYPES_TO_WAYOUT = <int, List<RoomInOut>> {
    0   : [],

    1   : [RoomInOut([Direction.top, Direction.right, Direction.left],
                     [Direction.bottom])],

    2   : [RoomInOut([Direction.right, Direction.left],
                     [Direction.right, Direction.left])],

    3   : [RoomInOut([Direction.top],
                     [Direction.bottom])],

    4   : [RoomInOut([Direction.top], [Direction.left]),
           RoomInOut([Direction.right], [Direction.bottom])],

    5   : [RoomInOut([Direction.top], [Direction.right]),
           RoomInOut([Direction.left], [Direction.bottom])],

    6   : [RoomInOut([Direction.right, Direction.left],
                     [Direction.right, Direction.left])],

    7   : [RoomInOut([Direction.top, Direction.right],
                     [Direction.bottom])],

    8   : [RoomInOut([Direction.right, Direction.left],
                     [Direction.bottom])],

    9   : [RoomInOut([Direction.top, Direction.left],
                     [Direction.bottom])],

    10  : [RoomInOut([Direction.top],
                     [Direction.left])],

    11  : [RoomInOut([Direction.top],
                     [Direction.right])],

    12  : [RoomInOut([Direction.right],
                     [Direction.bottom])],

    13  : [RoomInOut([Direction.left],
                     [Direction.bottom])],
};

class Pair {
  final int first;
  final int second;
  Pair(this.first, this.second){}
}

class IndiPos {
  final int _x;
  final int _y;
  final Direction _pos;

  IndiPos(this._x, this._y, this._pos) {}

  static IndiPos read() {
    final vInputs = readLineSync().split(' ');
    return IndiPos(int.parse(vInputs[0]),
                   int.parse(vInputs[1]),
                   Direction.read(vInputs[2]));
  }

  Pair    get xy    => Pair(_x, _y);
  String  get xyStr => "${_x} ${_y}";
    
  IndiPos move(Direction aDir) {
      return IndiPos(_x+aDir.dx, _y+aDir.dy, aDir.from);
  }
}



class Game {
  final int                   _w;
  final int                   _h;
  final List<List<RoomType>>  _lines;
  final int                   _ex;
  
  Game(this._w, this._h, this._lines, this._ex) {}

  RoomType roomAt(int aX, int aY) => _lines[aY][aX];


  void loop() {
    while (true) {
      final vIPos=IndiPos.read();
      logd("## x: ${vIPos._x}, y: ${vIPos._y}, pos: ${vIPos._pos.str}");
      print(action(vIPos));
    }
  }


  List<RoomInOut> possibleWayouts(int aRoom, Direction aPos) {
    String showPoss(List<Direction> aPoss) {
      return aPoss.map( (x) => x.str ).join(",");
      //  return aPoss.joinToString(transform={d: Direction -> d.str});
    }

    if (ROOM_TYPES_TO_WAYOUT.containsKey(aRoom)) {
        final vRoomsInOuts=ROOM_TYPES_TO_WAYOUT[aRoom]!;
        logd("# Room: ${aRoom}: ${vRoomsInOuts.length}");
        for (int i=0; i<vRoomsInOuts.length; ++i) {
          logd("#  ${i}: ${showPoss(vRoomsInOuts[i].ins)} -> ${showPoss(vRoomsInOuts[i].outs)}");
        }
        return vRoomsInOuts.where((x) => x._ins.contains(aPos)).toList();
    } else {
        throw Exception("Not a valid room type: ${aRoom}");
    }
  }


  String action(IndiPos aIPos) {
    final vRoom=roomAt(aIPos._x, aIPos._y);
    final vInOuts=possibleWayouts(vRoom, aIPos._pos);
    assert(vInOuts[0]._ins.contains(aIPos._pos));
    if (1 == vInOuts.length) {
      if(1 == vInOuts[0]._outs.length) {
        return aIPos.move(vInOuts[0]._outs[0]).xyStr;
      } else {
        // Exclude way in from way outs
        final vOuts=vInOuts[0]._outs.where((x) => x!=aIPos._pos).toList();
        if (1 == vOuts.length) {
          return aIPos.move(vOuts[0]).xyStr;
        } else {
          throw Exception("To be implemented 1");
        }
      }
    }else{
      throw Exception("To be implemented 2");
    }
  }

  static Game read() {
    List vInputs=readLineSync().split(' ');
    int vW = int.parse(vInputs[0]); // number of columns.
    int vH = int.parse(vInputs[1]); // number of rows.

    return Game(vW, vH,
                [for(var y=0; y<vH; ++y) readLineSync().split(" ").map( (x) => int.parse(x)).toList()],
                int.parse(readLineSync())); // the coordinate along the X axis of the exit (not useful for this first mission, but must be read).
  }

  Game logTunnel() {
    String format(int x) {
      return (x<10 ? " ": "")+x.toString();
    }

    logd("## Tunnel: ${_w}x${_h}");
    for (int iY=0; iY<_h; ++iY) {
      logd("## ${format(iY)}: ", false);
      for(int iX=0; iX<_w; ++iX) {
          logd("${format(_lines[iY][iX])} ", false);
      }
      logd("");
    }
    return this;
  }

}

void main() {
    Game vGame=Game.read();
    vGame.logTunnel();
    vGame.loop();
}

File: puzzles/The-Fall-Episode-1/the_fall_episode_1.py
import sys
import math
from collections import namedtuple

def logd(aS): print(aS, file=sys.stderr)


_Direction = namedtuple("_Direction", [
    "str", "dx", "dy", "from_"
])

class Direction(_Direction):
    def read(aS):
        if "TOP" == aS:
            return Direction.TOP
        elif "RIGHT" == aS:
            return Direction.RIGHT
        elif "BOTTOM" == aS:
            return Direction.BOTTOM
        elif "LEFT" == aS:
            return Direction.LEFT
        else:
            raise ValueError(f"Invalid direction: {aS}")

Direction.TOP       = Direction("TOP",    0,  -1, "BOTTOM")
Direction.RIGHT     = Direction("RIGHT",  +1, 0,  "LEFT")
Direction.BOTTOM    = Direction("BOTTOM", 0,  +1, "TOP")
Direction.LEFT      = Direction("LEFT",   -1, 0,  "RIGHT")


RoomInOut = namedtuple("RoomInOut", [
    "ins",  # :: [Direction]
    "outs", # :: [Direction]
])

ROOM_TYPES_TO_WAYOUT = {
    0   : [],

    1   : [RoomInOut([Direction.TOP, Direction.RIGHT, Direction.LEFT],
                     [Direction.BOTTOM])],

    2   : [RoomInOut([Direction.RIGHT, Direction.LEFT],
                     [Direction.RIGHT, Direction.LEFT])],

    3   : [RoomInOut([Direction.TOP],
                     [Direction.BOTTOM])],

    4   : [RoomInOut([Direction.TOP], [Direction.LEFT]),
           RoomInOut([Direction.RIGHT], [Direction.BOTTOM])],

    5   : [RoomInOut([Direction.TOP], [Direction.RIGHT]),
           RoomInOut([Direction.LEFT], [Direction.BOTTOM])],

    6   : [RoomInOut([Direction.RIGHT, Direction.LEFT],
                     [Direction.RIGHT, Direction.LEFT])],

    7   : [RoomInOut([Direction.TOP, Direction.RIGHT],
                     [Direction.BOTTOM])],

    8   : [RoomInOut([Direction.RIGHT, Direction.LEFT],
                     [Direction.BOTTOM])],

    9   : [RoomInOut([Direction.TOP, Direction.LEFT],
                     [Direction.BOTTOM])],

    10  : [RoomInOut([Direction.TOP],
                     [Direction.LEFT])],

    11  : [RoomInOut([Direction.TOP],
                     [Direction.RIGHT])],

    12  : [RoomInOut([Direction.RIGHT],
                     [Direction.BOTTOM])],

    13  : [RoomInOut([Direction.LEFT],
                     [Direction.BOTTOM])],
}


_IndiPos = namedtuple("_IndiPos", [
    "x", "y", "pos"
])

class IndiPos(_IndiPos):
    def read():
        v_inputs = input().split()
        return IndiPos(x    = int(v_inputs[0]),
                       y    = int(v_inputs[1]),
                       pos  = Direction.read(v_inputs[2]))

    @property
    def xy(self):
        return (self.x, self.y)

    @property
    def xy_str(self):
        return f"{self.x} {self.y}"
    
    def move(self, a_dir):
        return IndiPos(self.x+a_dir.dx, self.y+a_dir.dy, a_dir.from_)
    

_Game = namedtuple("_Game", [
    "w", "h", "lines", "ex"
])

class Game(_Game):
    def read():
        vW, vH = (int(x) for x in input().split())
        return Game(w       = vW,
                    h       = vH,
                    lines   = [ [int(x) for x in input().split()]
                                for _ in range(vH)
                              ],
                    ex      = int(input())
        )
    def room_at(self, a_x, a_y):
        return self.lines[a_y][a_x]

    def logTunnel(self):
        logd(f"## Tunnel: {self.w}x{self.h}")
        logd("\n".join(f"##  {n:2}: "+" ".join(f"{x:2}" for x in line)
                       for n, line in enumerate(self.lines)))
        return self

    def loop(self):
        while True:
            vIPos = IndiPos.read()
            logd(f"## x: {vIPos.x}, y: {vIPos.y}, pos: {vIPos.pos.str} (from: {vIPos.pos.from_})")
            print(self.action(vIPos));

    def action(self, a_ipos):
        v_room = self.room_at(a_ipos.x, a_ipos.y)
        v_inouts = self.possibleWayouts(v_room, a_ipos.pos)
        assert(a_ipos.pos in v_inouts[0].ins)
        if 1 == len(v_inouts):
            if 1 == len(v_inouts[0].outs):
                return a_ipos.move(v_inouts[0].outs[0]).xy_str
            else:
                # Exclude way in from way outs
                v_outs = list(filter(lambda x: x!=a_ipos.pos, v_inouts[0].outs))
                if 1 == len(v_outs):
                    return a_ipos.move(v_outs[0]).xy_str
                else:
                    raise Exception("To be implemented 1")
        else:
            raise Exception("To be implemented 2")


    def possibleWayouts(self, a_room, a_pos):
        """ Returns: [RoomInOut] """
        def show_poss(a_poss):
            return ", ".join(d.str for d in a_poss)

        if a_room in ROOM_TYPES_TO_WAYOUT:
            v_rooms_inouts = ROOM_TYPES_TO_WAYOUT[a_room]
            logd("# Room: {a_room}: {len(v_rooms_inouts)}")
            for i in range(len(v_rooms_inouts)):
                logd(f"#  {i}: {show_poss(v_rooms_inouts[i].ins)} -> {show_poss(v_rooms_inouts[i].outs)}")
            return list(filter(lambda x: a_pos in x.ins, v_rooms_inouts))
        else:
            raise ValueError("Not a valid room type: {a_room}")



# Main
(
    Game.read()
        .logTunnel()
        .loop()
)
